### PR TITLE
[STACK-1321, 1322]: Added delete_member and delete_hm flow in delete_pool flow

### PR DESCRIPTION
**Pull request by afrin-chakure-a10**
## Description
- **Severity Level**: Critical
- Health monitor configuration was not getting removed from vThunder when we delete its associated Pool from openstack, even though, db entry for `health_monitor` gets removed.
- Delete a pool that has a member inside, will succeed in Octavia. But on Thunder, the server(member) is still there.

## Jira Ticket
https://a10networks.atlassian.net/browse/STACK-1321
https://a10networks.atlassian.net/browse/STACK-1322

## Technical Approach
- When pool is deleted, the child `members` and `hm` are deleted before.
- Invoke `delete_member_vthunder_subflow` flow inside `delete_pool_flow` flow in `a10_pool_flows.py`, if members are present.
- If the very last member of the project is deleted, the `delete_vrid_subflow` is also invoked. This is taken care by `mem_count` in 
`controller_worker` module.
- Invoke `delete_health_monitor_vthunder_subflow` flow inside `delete_pool` flow, if there exist an hm associated with that pool.
- Since the flow for `pool`, `member` and `hm` is same for `hardware_thunder` and `automated vthunder flows`, it will work in similar way.

## Config Changes
None

## Test Cases
1) **Normal Env**
- When pool is deleted via openstack cli command, the `health_monitor` associated with it are also deleted from vthunder config.
- When pool is deleted via openstack cli command, the `members` are also deleted from vthunder.
- Quotas in `quotas` table is correct before, during and after cli commands for `in_use_member`, `in_use_pool` and `in_use_health_monitor`.

2) **VRID config Env**
Config settings eg:
```
[a10_global]
vrid_floating_ip = 'dhcp'
```

- When pool is deleted via openstack cli command, the `health_monitor` associated with it are also deleted from vthunder config.
- When pool is deleted via openstack cli command, the `members` are also deleted from vthunder.
- VRID configuration corresponding to these members, is also deleted configuration from vthunder, in case the very last member in the `pool.project_id` is also deleted.
- Quotas in `quotas` table is correct before, during and after cli commands  for `in_use_member`, `in_use_pool` and `in_use_health_monitor`.

3) **VLAN config Env**
Config settings eg:
```
[a10_global]
network_type= 'vlan'

[hardware_thunder]
 devices=[
 {
 "project_id": "<>",
 "ip_address": "10.0.0.99",
 "device_name": "rack_vthunder",
 "username": "<>",
 "password": "<>",
 "interface_vlan_map": {
  "device_1": {
    "vcs_device_id": 1,
    "mgmt_ip_address": "10.0.0.74",
      "ethernet_interfaces": [{
        "interface_num": 2,
          "vlan_map": [{"vlan_id": 12, "ve_ip" : ".15"}]
   }],
   "trunk_interfaces": [{
    "interface_num": 1,
    "vlan_map": [
     {"vlan_id": 13, "use_dhcp": "True"},
     {"vlan_id": 14, "use_dhcp": "True"}
    ]
   }],
  },
<<similar device 2 config>>
```
- When pool is deleted via openstack cli command, the `health_monitor` associated with it are also deleted from vthunder config.
- When pool is deleted via openstack cli command, the `members` are also deleted from vthunder.
- VLAN interfaces corresponding to these members, are also deleted with ve interfaces configuration on vthunder.
- Quotas in `quotas` table is correct before, during and after cli commands  for `in_use_member`, `in_use_pool` and `in_use_health_monitor`.

## Manual Testing

In a complete SLB tree created environment, following scenarios occur in different environment:

1) **Normal Environment**
- `openstack loadbalancer pool delete pool1`
- Run `show running-config` in vthunder cli
- Check on vThunder, along with `service-group`, associated `health_monitor` is deleted
- On vthunder, all the associated `servers` are deleted.
- Check mysql table : `select * from quotas;` and check  `in_use_member`, `in_use_pool` and `in_use_health_monitor` column values.

2) **With VRID configured Environment**
- `openstack loadbalancer pool delete pool1`
- Run `show running-config` in vthunder cli
- Check on vThunder, along with `service-group`, associated `health_monitor` is deleted
- On vthunder, all the associated `servers` are deleted.
- `vrid` config is also deleted, if the very last `server` associated with that project is deleted.
- Check mysql table : `select * from quotas;` and check  `in_use_member`, `in_use_pool` and `in_use_health_monitor` column values.

3) **With VLAN configured environment**
- `openstack loadbalancer pool delete pool1`
- Run `show running-config` in vthunder cli
- Check on vThunder, `service-group` is deleted.
-  The associated `health_monitor` is also deleted.
- All the associated `server`'s are deleted.
- All Member related `vlan interface`  and `interface ve` are also deleted.
- Check mysql table : `select * from quotas;` and check  `in_use_member`, `in_use_pool` and `in_use_health_monitor` column values.

## Scripts for testing
```
openstack loadbalancer create --vip-subnet-id 91e9b009-4b7c-421b-b5ee-646fc86187db --name SLB1
sleep 6
openstack loadbalancer listener create  --protocol HTTP --protocol-port 80 SLB1 --name listener1
openstack loadbalancer pool create  --protocol HTTP --listener listener1 --lb-algorithm ROUND_ROBIN --name pool1
openstack loadbalancer member create --address 10.0.13.120 --subnet-id 2e3662b5-13bd-4f2c-a6dd-6006878dd791 --protocol-port 80 pool1 --name member1
sleep 6
openstack loadbalancer member create --address 10.0.13.124 --subnet-id 2e3662b5-13bd-4f2c-a6dd-6006878dd791 --protocol-port 80 pool1 --name member2
```

### CODE REVIEW

File: a10_octavia/controller/worker/flows/a10_member_flows.py [hunk 1]
```diff
@@ -129,31 +129,79 @@ def get_delete_member_flow(self):
...
                                              constants.LISTENERS]))
-        delete_member_flow.add(a10_database_tasks.GetVRIDForProjectMember(
+        delete_member_flow.add(self.get_delete_member_vrid_subflow())
+        delete_member_flow.add(vthunder_tasks.WriteMemory(
+            requires=a10constants.VTHUNDER))
+        return delete_member_flow
+
+    def get_delete_member_vthunder_subflow(self):
+        delete_member_vthunder_subflow = linear_flow.Flow(
+            a10constants.DELETE_MEMBER_VTHUNDER_SUBFLOW)
```
Comment (hthompson-a10) [inline a10_octavia/controller/worker/flows/a10_member_flows.py:148]: `get_delete_member_vthunder_subflow` and `get_delete_member_vthunder_internal_subflow` shouldn't have the same subflow name

File: a10_octavia/controller/worker/flows/a10_member_flows.py [hunk 2]
```diff
@@ -129,31 +129,79 @@ def get_delete_member_flow(self):
...
+            delete_member_thunder_subflow.add(vthunder_tasks.DeleteInterfaceTagIfNotInUseForMember(
+                name='delete_unused_interface_tag_in_member_' + member_id,
+                requires=[constants.MEMBER, a10constants.VTHUNDER],
+                rebind={constants.MEMBER: member_id}))
+
+        return delete_member_thunder_subflow
+
+    def get_delete_member_vrid_subflow(self):
+        delete_member_vrid_subflow = linear_flow.Flow(
+            a10constants.DELETE_MEMBER_VRID_SUBFLOW)
```
Comment (hthompson-a10) [inline a10_octavia/controller/worker/flows/a10_member_flows.py:177]: `get_delete_member_vrid_subflow` and `a10constants.DELETE_MEMBER_VRID_SUBFLOW` shouldn't have the same subflow name.

File: a10_octavia/controller/worker/flows/a10_pool_flows.py [hunk 3]
```diff
@@ -97,6 +106,26 @@ def get_delete_pool_flow(self):
...
 
+    def _get_delete_health_monitor_vthunder_subflow(self, health_mon):
+        delete_hm_vthunder_subflow = linear_flow.Flow('hm_delete_subflow_with_pool')
+        if health_mon:
+            delete_hm_vthunder_subflow.add(
+                self.hm_flow.get_delete_health_monitor_vthunder_subflow())
+        return delete_hm_vthunder_subflow
+
+    def _get_delete_member_vthunder_subflow(self, members, store):
+        delete_member_vthunder_subflow = linear_flow.Flow('members_delete_subflow_with_pool')
```
Comment (hthompson-a10) [inline a10_octavia/controller/worker/flows/a10_pool_flows.py:117]: `'members_delete_subflow_with_pool` needs to be turned into a constant

File: a10_octavia/controller/worker/flows/a10_pool_flows.py [hunk 4]
```diff
@@ -97,6 +106,26 @@ def get_delete_pool_flow(self):
 
         return delete_pool_flow
 
+    def _get_delete_health_monitor_vthunder_subflow(self, health_mon):
+        delete_hm_vthunder_subflow = linear_flow.Flow('hm_delete_subflow_with_pool')
```
Comment (hthompson-a10) [inline a10_octavia/controller/worker/flows/a10_pool_flows.py:110]: `'hm_delete_subflow_with_pool'` needs to be turned into a constant

**Comment by afrin-chakure-a10** [comment]
> Please let me know output of below scenario:
> 
> create a LB, listener, pool, memebers, hm ... try to delete pool but add a manual raise in member delete code, so it fails.
> Figure out status of members and pool and wether reverts are called properly.

Tried above scenario and seems reverts are not working properly as there are no reverts on member delete task
- Pool goes in `PENDING_DELETE` state
- Members in `ACTIVE` state
- Loadbalancer in `PENDING_UPDATE` state

on vthunder : Nothing gets deleted.

**Comment by hthompson-a10** [comment]
This error will be fixed when the PR #129 is merged then

**Pull request merged**